src/Overcrowdin/CrowdInHelperBase.cs

```diff
@@ -0,0 +1,326 @@
+					_existingDirectories.Add(directory);
+					parentDir = directory.Id;
+				}
+			}
+
+			// Create new storage for the file and upload it.
+			StorageResource storage;
+			using (Stream stream = new MemoryStream(Encoding.UTF8.GetBytes(fileData), false))
+				storage = await _client.Storage.AddStorage(stream, fileName);
+
+			var directoryId = directory?.Id ?? 0; // Files can't have a null directory or branch IDs for some reason
+			var branchSearchId = _branchId ?? 0;
+			var existingFile = _existingFiles.Find(f =>
+				f.BranchId == branchSearchId && f.DirectoryId == directoryId && string.Equals(f.Name, fileName, StringComparison.OrdinalIgnoreCase));
+			if (existingFile != null)
+			{
+				// File already exists so move it from storage onto the existing file, updating it
+				var request = new ReplaceFileRequest
+				{
+					StorageId = storage.Id,
+					UpdateOption = FileUpdateOption.KeepTranslationsAndApprovals
+				};
+
+				ValueTuple<File, bool?> response = await _executor.UpdateOrRestoreFile(_project.Id, existingFile.Id, request);
+				_existingFiles.Remove(existingFile);
+				Console.WriteLine($"Updated file {response.Item1.Path} ({(response.Item2 == true ? "modified" : "no change")})");
+			}
+			else
+			{
+				// File does not exist so move it from storage as a new file
+				var request = new AddFileRequest
+				{
+					Name = fileName,
+					StorageId = storage.Id,
+					BranchId = _branchId,
+					DirectoryId = directory?.Id,
+					Type = fileType
+				};
+
+				// appears that both values can't be null, so try using 0 for directory id
+				if (request.BranchId == null && request.DirectoryId == null)
+					request.DirectoryId = 0;
+
+				var newFile = await _executor.AddFile(_project.Id, request);
+				Console.WriteLine($"Added file {newFile.Path}");
+			}
+
+			// Delete storage that was created for the file (no need to keep it around)
+			await _client.Storage.DeleteStorage(storage.Id);
+		}
+
+		#endregion
+
+		#region Private helper methods
+		private async Task<int?> BuildTranslation(List<string> languages, bool skipUntranslated)
+		{
+			Console.WriteLine($"Starting translation build for {string.Join(", ", languages)}...");
+
+			BuildProjectTranslationRequest request = new TranslationCreateProjectBuildForm
+			{
+				BranchId = _branchId,
+				TargetLanguageIds = languages,
+				SkipUntranslatedFiles = skipUntranslated
+			};
+
+			ProjectBuild build = await _client.Translations.BuildProjectTranslation(_project.Id, request);
+			if (build.Status == BuildStatus.InProgress)
+				build = await WaitForBuildToComplete(build);
+
+			if (build.Status == BuildStatus.Failed || build.Status == BuildStatus.Canceled)
+			{
+				Console.Error.WriteLine("Translation build stopped unexpectedly: " + build.Status);
+				return null;
+			}
+
+			Console.WriteLine("Translation build finished");
+			return build.Id;
+		}
+
+		private async Task<ProjectBuild> WaitForBuildToComplete(ProjectBuild build)
```

Comment:
there should probably be some kind of timeout for this check, at the moment it looks like it'll just go forever.